program/c/src/oracle/model/price_model.c

```diff
@@ -1,99 +1,108 @@
 
-#define SORT_NAME  int64_sort_ascending
-#define SORT_KEY_T int64_t
-#include "../sort/tmpl/sort_stable.c"
+/*
+ * Quick select implementation highly optimized for reducing the number of compute units in BPF.
+ *
+ * On Floyd and Rivest's SELECT algorithm" by Krzysztof C. Kiwiel article
+ * mentions that the average case of this algorithm is is 2.5n + o(n) comparisons.
+ * The article itself presents a different algorithm (Floyd and Rivest) which
+ * performs faster in larger arrays but has a higher overhead. Other determinstic
+ * algorithms like median of medians have a higher overhead (or comibnations of
+ * them and quick select) and are not suitable for BPF.
+ */
+int64_t quick_select( int64_t * a, uint64_t n, uint64_t k ) {
+    while (1) {
+        if (n == 1) return a[0];
+
+        int64_t pivot;
+
+        /*
+         * Select median of 3 as pivot which massively reduces the number of
+         * comparisons in the average case and also helps in reducing the likelihood
+         * of worst case time complexity.
+         */
+        {
+            if (a[0] < a[n / 2]) {
+                if (a[n / 2] < a[n - 1]) pivot = a[n / 2];
+                else if (a[0] < a[n - 1]) pivot = a[n - 1];
+                else pivot = a[0];
+            } else {
+                if (a[n / 2] > a[n - 1]) pivot = a[n / 2];
+                else if (a[0] > a[n - 1]) pivot = a[n - 1];
+                else pivot = a[0];
+            }
+        }
+
+        /*
+         * Partition the array around the pivot. This partitioning is done in a
+         * way that if the array has a lot of duplicate pivots, it doesn't result in
+         * quadratic time complexity and evenly distributes the pivots.
+         *
+         * When the loop is done, the array looks like any of the following:
+         * 1. ....ji....
+         * In this case the elements a[x] for x <= j are <= pivot and a[x] for x >= i are >= pivot.
+         *
+         * 2. ....j.i...
+         * In this case same conditions as above apply but a[j+1] == pivot because the last step
+         * resulting in j and i in this position is i==j and swapping which means that that element
+         * is neither < pivot nor > pivot.
+         *
+         * In the case of multiple pivot elements, the elements will be evenly distributed because
+         * i nor j won't move if the element is equal to the pivot and if both be equal to the pivot
+         * they will swap (which is ineffective) and move forward which results in one pivot to the
+         * left and one to the right.
+         */
+        uint64_t i = 0, j = n - 1;
+        while (i <= j) {
+            if (a[i] < pivot) {
+                i++;
+                continue;
+            }
+            if (a[j] > pivot) {
+                j--;
+                continue;
+            }
+
+            int64_t tmp = a[i];
+            a[i] = a[j];
+            a[j] = tmp;
+            i++;
+            j--;
+        }
```

Comment:
This is potentially overly nitpicking but you can reduce the instruction count slightly here:

```
while (i <= j) {
    if (a[i] < pivot) {
        i++;
    } else if (a[j] > pivot) {
        j--;
    } else {
        int64_t tmp = a[i];
        a[i] = a[j];
        a[j] = tmp;
        i++;
        j--;
    }
}
```

With the continue statements you force a loop condition when the loop could have exited, so the `continue` statements force a recheck of the while condition unnecessarily, I doubt it is going to matter that much compared to the improvements you've made but just mentioning, the if/else form might be more readable though.

Reply:
Same here no benefit, if the compiler can't optimize this I'd be shocked and even if it did it's only two instructions. Ignore.

Reply:
I checked it, the result was the same. Probably compiler does exactly what you said.